src/main/java/de/tum/cit/aet/artemis/lti/service/Lti13Service.java

```diff
@@ -311,6 +335,178 @@ private Optional<Exercise> getExerciseFromTargetLink(String targetLinkUrl) {
+
+    /**
+     * Retrieves an Optional of a Course referenced by the given targetLinkUrl.
+     * This method extracts the course ID from the URL using predefined path patterns
+     * and then fetches the corresponding Course from the repository.
+     * Returns an empty Optional in the following cases:
+     * - The targetLinkUrl is malformed and cannot be parsed.
+     * - The URL does not match any known path pattern.
+     * - The extracted path does not contain a "courseId" variable.
+     * - No Course with the extracted ID exists in the repository.
+     *
+     * @param targetLinkUrl the target link URL to retrieve a Course
+     * @return an Optional containing the Course if found, or an empty Optional otherwise
+     */
+    public Optional<Course> getCourseFromTargetLink(String targetLinkUrl) {
+        AntPathMatcher matcher = new AntPathMatcher();
+
+        String targetLinkPath;
+        try {
+            targetLinkPath = new URI(targetLinkUrl).getPath();
+        }
+        catch (URISyntaxException ex) {
+            log.info("Malformed target link url: {}", targetLinkUrl);
+            return Optional.empty();
+        }
+
+        Map<String, String> pathVariables = null;
+        if (matcher.match(COURSE_PATH_PATTERN, targetLinkPath)) {
+            pathVariables = matcher.extractUriTemplateVariables(COURSE_PATH_PATTERN, targetLinkPath);
+        }
+        else if (matcher.match(COMPETENCY_PATH_PATTERN, targetLinkPath)) {
+            pathVariables = matcher.extractUriTemplateVariables(COMPETENCY_PATH_PATTERN, targetLinkPath);
+        }
+        else if (matcher.match(EXERCISE_PATH_PATTERN, targetLinkPath)) {
+            pathVariables = matcher.extractUriTemplateVariables(EXERCISE_PATH_PATTERN, targetLinkPath);
+        }
+        else if (matcher.match(LEARNING_PATH_PATH_PATTERN, targetLinkPath)) {
+            pathVariables = matcher.extractUriTemplateVariables(LEARNING_PATH_PATH_PATTERN, targetLinkPath);
+        }
+        else if (matcher.match(LECTURE_PATH_PATTERN, targetLinkPath)) {
+            pathVariables = matcher.extractUriTemplateVariables(LECTURE_PATH_PATTERN, targetLinkPath);
+        }
+        else if (matcher.match(IRIS_PATH_PATTERN, targetLinkPath)) {
+            pathVariables = matcher.extractUriTemplateVariables(IRIS_PATH_PATTERN, targetLinkPath);
+        }
+
+        if (pathVariables == null || !pathVariables.containsKey("courseId")) {
+            log.info("Could not extract courseId from target link: {}", targetLinkUrl);
+            return Optional.empty();
+        }
+
+        String courseId = pathVariables.get("courseId");
+        return courseRepository.findById(Long.valueOf(courseId));
+    }
+
+    /**
+     * Checks if the target link URL references a competency.
+     *
+     * @param targetLinkUrl the target link URL
+     * @return true if the target link URL references a competency, false otherwise
+     */
+    private boolean targetLinkHasCompetency(String targetLinkUrl) {
```

Comment:
`targetLinkHasCompetency(...)`, `targetLinkHasLearningPath(...)` and `targetLinkHasIris(...)` are almost identical. This is really bad code. I kindly ask you to unify the code already here to avoid such issues. You could e.g. have one method that returns an enum type whether competency, learningPath or iris is included.